src/lib/common/mtmodule.py

```diff
@@ -41,7 +95,81 @@ def wrapper(self, *args):
 
         return decorator
 
+    @staticmethod
+    def batched_phase(phase_key, remove_db=True):
+        """
+        Run a phase in parallel using multiprocessing. Can only be applied to a class function that takes a single argument that is of GeneratorType.
+        """
+
+        def decorator(innards):
+            @wraps(innards)
+            def wrapper(self, *args):
+                if not isinstance(self, MTModule):
+                    raise ImproperLoggedPhaseError(innards.__name__)
+                if len(args) < 1 or not isinstance(args[0], GeneratorType):
+                    raise BatchedPhaseArgNotGenerator(innards.__name__)
+
+                self.PHASE_KEY = phase_key
+
+                all_elements = list(args[0])
+                batch_size = get_batch_size(len(all_elements))
+                other_args = args[1:]
+                # each chunk is a generator
+                cs = batch(all_elements, n=batch_size)
+                
+                manager = multiprocessing.Manager()
+
+                # switch logs to multiprocess access list
+                self.__LOGS = manager.list()
+                done_queue = manager.Queue()
+                batches_running = manager.Value('i', 1)
+
+                dbfile = f"{self.BASE_DIR}/{self.UNIQUE_ID}.db"
+
+                done_dict = {}
+                try:
+                    with open(dbfile, 'rb') as f:
+                        _bytes = f.read(8)
+                        while _bytes:
+                            entry = struct.unpack("II", _bytes)
+                            if entry[0] not in done_dict:
+                                done_dict[entry[0]] = {}
+                            done_dict[entry[0]][entry[1]] = 1 
+                            _bytes = f.read(8)
+                        f.close()
+                except:
```

Comment:
@ivansafrin what are the fail cases here? Should we raise a descriptive error to say something happened that shouldn't have?

Reply:
It's a pass currently because that's the expected case of the progress database file not existing unless the job was previously interrupted. A more robust way to error check would be to see if the file actually exists first, then the failure to open it would indicate an actual IO error. Up to you if you think it should be that granular.